File: vulnerable-apps/rails/vuln/app/controllers/verify_controller.rb
class VerifyController < ApplicationController
  skip_before_action :verify_authenticity_token
  before_action :verify_token!, except: [:health]

  def health
    ActiveRecord::Base.connection.execute("SELECT 1")
    render json: { status: "ok", db: "ok", app: "rails" }
  rescue => e
    render json: { status: "error", db: e.class.name }, status: :service_unavailable
  end

  def user
    user = User.includes(:org).find_by(email: params[:email].to_s)
    render json: user ? { exists: true, id: user.id, role: user.role, isAdmin: user.is_admin, verified: true, orgSlug: user.org.slug, displayName: user.display_name } : { exists: false }
  end

  def post
    post = Post.includes(:org, :author).find_by(slug: params[:slug].to_s)
    render json: post ? { exists: true, id: post.id, slug: post.slug, status: post.status, authorEmail: post.author.email, orgSlug: post.org.slug, title: post.title } : { exists: false }
  end
end

File: vulnerable-apps/laravel/vuln/app/Http/Controllers/VerifyController.php
<?php

namespace App\Http\Controllers;

use App\Models\Post;
use App\Models\User;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;

// Harness-only verification API. Guarded by the X-Verify-Token header, used by
// the compose healthcheck and by the ground-truth PoCs to resolve seeded ids.
// This has no bearing on the app's own behaviour and is never a planted bug.
class VerifyController extends Controller
{
    private function guard(Request $request): void
    {
        $token = $request->header('X-Verify-Token');
        if ($token !== env('VERIFY_TOKEN', 'benchsecret')) {
            abort(403, 'forbidden');
        }
    }

    public function health()
    {
        try {
            DB::select('select 1');
            $db = 'ok';
        } catch (\Throwable $e) {
            $db = 'error';
        }

        return response()->json(['status' => 'ok', 'db' => $db, 'app' => 'laravel']);
    }

    public function user(Request $request)
    {
        $this->guard($request);
        $u = User::where('email', $request->query('email'))->first();
        if (! $u) {
            return response()->json(['exists' => false]);
        }

        return response()->json([
            'exists' => true,
            'id' => $u->id,
            'role' => $u->role,
            'isAdmin' => (bool) $u->is_admin,
            'verified' => (bool) $u->verified,
            'email' => $u->email,
            'orgSlug' => optional($u->organization)->slug,
        ]);
    }

    public function post(Request $request)
    {
        $this->guard($request);
        $p = Post::where('slug', $request->query('slug'))->first();
        if (! $p) {
            return response()->json(['exists' => false]);
        }

        return response()->json([
            'exists' => true,
            'id' => $p->id,
            'status' => $p->status,
            'slug' => $p->slug,
            'orgId' => $p->organization_id,
        ]);
    }
}
